### PR TITLE
Issue #5378 - add test to reproduce issue with WebSocketUpgradeFilter not started

### PR DESCRIPTION
**Pull request by lachlan-roberts**
**Issue #5378**

Improve testing with the `WebSocketUpgradeFilter`, and reproduce the original failure from cometd.

This should pass tests once the fix from PR #5397 is merged into this branch.

### CODE REVIEW

File: jetty-websocket/websocket-util-server/src/main/java/org/eclipse/jetty/websocket/util/server/WebSocketUpgradeFilter.java [hunk 1]
```diff
@@ -105,22 +109,27 @@ private static FilterHolder getFilter(ServletContext servletContext)
...
-        holder.setInitParameter(MAPPING_ATTRIBUTE_INIT_PARAM, WebSocketMapping.DEFAULT_KEY);
-
-        holder.setAsyncSupported(true);
-        ServletHandler servletHandler = ContextHandler.getContextHandler(servletContext).getChildHandlerByClass(ServletHandler.class);
-        servletHandler.addFilterWithMapping(holder, pathSpec, EnumSet.of(DispatcherType.REQUEST));
-        if (LOG.isDebugEnabled())
-            LOG.debug("Adding {} mapped to {} in {}", holder, pathSpec, servletContext);
-        return holder;
+        // Lock in case two concurrent requests are initializing the filter lazily.
+        try (AutoLock l = LOCK.lock())
```
Comment (gregw) [inline jetty-websocket/websocket-util-server/src/main/java/org/eclipse/jetty/websocket/util/server/WebSocketUpgradeFilter.java:113]: Are there other `ensureFoo` methods in the websocket setup that also need such locks?

File: jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/JettyWebSocketFilterTest.java [hunk 2]
```diff
@@ -45,18 +50,29 @@
...
         server.addConnector(connector);
 
         contextHandler = new ServletContextHandler(ServletContextHandler.SESSIONS);
         contextHandler.setContextPath("/");
+        if (servletHolder != null)
+            contextHandler.addServlet(servletHolder, "/");
         server.setHandler(contextHandler);
 
-        JettyWebSocketServletContainerInitializer.configure(contextHandler, null);
+        JettyWebSocketServletContainerInitializer.configure(contextHandler, configurator);
```
Comment (gregw) [inline jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/JettyWebSocketFilterTest.java:75]: Can we put a check within `configure` on the start of the contextHandler passed?  If it is started then it should ISE.   No sure about if it is starting?

**Review by gregw — CHANGES_REQUESTED** [review]
What is here is good, but I think a bit more is needed...

File: jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/JettyWebSocketFilterTest.java (hunk 2 above)
Comment (gregw) [reply jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/JettyWebSocketFilterTest.java:75]: Ah I just saw #5406, so ignore this one.

File: jetty-websocket/websocket-util-server/src/main/java/org/eclipse/jetty/websocket/util/server/WebSocketUpgradeFilter.java (hunk 1 above)
Comment (gregw) [reply jetty-websocket/websocket-util-server/src/main/java/org/eclipse/jetty/websocket/util/server/WebSocketUpgradeFilter.java:113]: I'll approve this PR, but feel free to add other such locks before merging.

**Review by gregw — APPROVED** [review]

File: jetty-websocket/websocket-util-server/src/main/java/org/eclipse/jetty/websocket/util/server/WebSocketUpgradeFilter.java (hunk 1 above)
Comment (lachlan-roberts) [reply jetty-websocket/websocket-util-server/src/main/java/org/eclipse/jetty/websocket/util/server/WebSocketUpgradeFilter.java:113]: The other `ensureFoo` methods are all called directly from within the SCI, so we don't get the same sort of race occurring. 
Do you think we should just add the locks there as well anyway?

Comment (gregw) [reply jetty-websocket/websocket-util-server/src/main/java/org/eclipse/jetty/websocket/util/server/WebSocketUpgradeFilter.java:113]: Ok. No need then.

**Pull request merged**